File: Python/binary_search.py
def binary_search(arr: list[int], target: int) -> int | None:
    start = 0
    end = len(arr) - 1
    while start <=end:
        mid = (start + end) // 2
        if(arr[mid] == target):
            return mid
        if(target < arr[mid]):
            end = mid - 1
        else:
            start = mid + 1
    return None

def recursive_binary_search(arr: list[int], target: int) -> bool:
    if(len(arr) == 0):
        return False
    
    mid = len(arr) // 2
    if(target == arr[mid]):
        return True
    if(target < arr[mid]):
        return recursive_binary_search(arr[:mid], target)
    else:
        return recursive_binary_search(arr[mid + 1:], target)
    
def main():
    l = [0,1,2,3,4,5,6,7,8,9,10,11,12]
    print(f'Binary Search(7): {binary_search(l,7)}')
    print(f'Recursive Binary Search(7): {recursive_binary_search(l,7)}')

main()

File: dsa/binary_search.cpp
#include <iostream>
using namespace std;

bool recursive_binary_search(int, int[], int);
int *create_sorted_array(int);

int main() {
  int size = 0;
  cout << "Enter the size of array to create: ";
  cin >> size;
  int *arr = create_sorted_array(size);
  int target;
  cout << "\nEnter the number to search: ";
  cin >> target;
  cout << "\n"
       << target << " Found: " << recursive_binary_search(target, arr, size)
       << endl;
  return 0;
}

bool recursive_binary_search(int target, int arr[], int a_len) {
  if (a_len == 0) {
    return false;
  }

  int mid = a_len / 2;
  if (arr[mid - 1] == target) {
    return true;
  }
  if (target < arr[mid - 1]) {
    int *arr_2 = (int *)calloc(mid, sizeof(int));
    for (int i = 0; i < mid; i++) {
      arr_2[i] = arr[i];
    }
    return recursive_binary_search(target, arr_2, mid);
  }
  if (target > arr[mid - 1]) {
    int a2_len = a_len - mid + 1;
    int *arr_2 = (int *)calloc(a2_len, sizeof(int));
    for (int i = 0; i < a2_len; i++) {
      arr_2[i] = arr[a2_len + i];
    }
    return recursive_binary_search(target, arr_2, a2_len);
  }
  return false;
}

int *create_sorted_array(int n) {
  int *arr = (int *)calloc(n, sizeof(int));
  for (int i = 0; i < n; i++) {
    arr[i] = i;
  }
  return arr;
}
